src/Core/FakeVimProxy.cpp

```diff
@@ -0,0 +1,495 @@
+        }
+        else
+        {
+            return;
+        }
+    }
+    Q_ASSERT(doc);
+
+    const int indentSize = FakeVim::Internal::theFakeVimSetting(FakeVim::Internal::ConfigShiftWidth)->value().toInt();
+
+    QTextBlock startBlock = doc->findBlockByNumber(beginBlock);
+
+    // Record line lengths for mark adjustments
+    QVector<int> lineLengths(endBlock - beginBlock + 1);
+    QTextBlock block = startBlock;
+
+    for (int i = beginBlock; i <= endBlock; ++i)
+    {
+        const auto line = block.text();
+        lineLengths[i - beginBlock] = line.length();
+        if (typedChar.unicode() == 0 && line.simplified().isEmpty())
+        {
+            // clear empty lines
+            QTextCursor cursor(block);
+            while (!cursor.atBlockEnd())
+                cursor.deleteChar();
+        }
+        else
+        {
+            const auto previousBlock = block.previous();
+            const auto previousLine = previousBlock.isValid() ? previousBlock.text() : QString();
+
+            int indent = firstNonSpace(previousLine);
+            if (typedChar == '}')
+                indent = std::max(0, indent - indentSize);
+            else if (previousLine.endsWith("{"))
+                indent += indentSize;
+            const auto indentString = QString(" ").repeated(indent);
+
+            QTextCursor cursor(block);
+            cursor.beginEditBlock();
+            cursor.movePosition(QTextCursor::StartOfBlock);
+            cursor.movePosition(QTextCursor::NextCharacter, QTextCursor::KeepAnchor, firstNonSpace(line));
+            cursor.removeSelectedText();
+            cursor.insertText(indentString);
+            cursor.endEditBlock();
+        }
+        block = block.next();
+    }
+}
+
+void FakeVimProxy::checkForElectricCharacter(bool *result, QChar c)
+{
+    *result = c == '{' || c == '}';
+}
+
+int FakeVimProxy::firstNonSpace(const QString &text)
+{
+    int indent = 0;
+    while (indent < text.length() && text.at(indent) == ' ')
+        ++indent;
+    return indent;
+}
+
+void FakeVimProxy::updateExtraSelections()
+{
+    QTextEdit *editor = qobject_cast<QTextEdit *>(m_widget);
+    QPlainTextEdit *plainEditor = qobject_cast<QPlainTextEdit *>(m_widget);
+    if (editor)
+    {
+        editor->setExtraSelections(m_clearSelection + m_searchSelection + m_blockSelection);
+    }
+    else if (plainEditor)
+    {
+        plainEditor->setExtraSelections(m_clearSelection + m_searchSelection + m_blockSelection);
+    }
+}
+
+bool FakeVimProxy::wantSaveAndQuit(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.cmd == "wq";
+}
+
+bool FakeVimProxy::wantSave(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.matches("w", "write") || cmd.matches("wa", "wall");
+}
+
+bool FakeVimProxy::wantQuit(const FakeVim::Internal::ExCommand &cmd)
+{
+    return cmd.matches("q", "quit") || cmd.matches("qa", "qall");
+}
+
+bool FakeVimProxy::save()
+{
+    if (!hasChanges())
+        return true;
```

Comment:
```suggestion
```

This is incorrect when the tab is untitled and the content is the same as the template.

You don't need to care about this, `MainWindow::save` checks everything.

Besides the mistake and duplicate code, it checks twice, which affects the performance.